src/app/clusters/network-commissioning/network-commissioning.cpp

```diff
@@ -0,0 +1,253 @@
+#include <platform/ConnectivityManager.h>
+#include <transport/NetworkProvisioning.h>
+
+#if defined(CHIP_DEVICE_LAYER_TARGET)
+#define DEVICENETWORKPROVISIONING_HEADER <platform/CHIP_DEVICE_LAYER_TARGET/DeviceNetworkProvisioningDelegateImpl.h>
+#include DEVICENETWORKPROVISIONING_HEADER
+#endif
+
+namespace chip {
+namespace app {
+namespace clusters {
+namespace NetworkCommissioning {
+
+constexpr uint8_t kMaxNetworkIDLen       = 32;
+constexpr uint8_t kMaxThreadDatasetLen   = 128;
+constexpr uint8_t kMaxWiFiSSIDLen        = 32;
+constexpr uint8_t kMaxWiFiCredentialsLen = 64;
+constexpr uint8_t kMaxNetworks           = 4;
+constexpr char kThreadNetworkName[]      = "tmp-thread-network";
+
+// TODO: The network info should be stored in attribute storage instead of memory.
+// TODO: We cannot get length of operational dataset here, should be fixed after we have byte string type.
+// TODO: Should encode response for command responses.
+enum class NetworkCommissioninError : uint8_t
+{
+    kSuccess                = 0,
+    kOutOfRange             = 1,
+    kBoundsExceeded         = 2,
+    kNetworkIDNotFound      = 3,
+    kDuplicateNetworkID     = 4,
+    kNetworkNotFound        = 5,
+    kRegulatoryError        = 6,
+    kAuthFailure            = 7,
+    kUnsupportedSecurity    = 8,
+    kOtherConnectionFailure = 9,
+    kIPV6Failed             = 10,
+    kIPBindFailed           = 11,
+    kLabel9                 = 12,
+    kLabel10                = 13,
+    kLabel11                = 14,
+    kLabel12                = 15,
+    kLabel13                = 16,
+    kLabel14                = 17,
+    kLabel15                = 18,
+    kUnknownError           = 19,
+};
+
+enum class NetworkType : uint8_t
+{
+    kUndefined = 0,
+    kWiFi      = 1,
+    kThread    = 2,
+    kEthernet  = 3,
+};
+
+enum class WiFiSecurityType : uint8_t
+{
+    kUnencrypted  = 0,
+    kWEPPersonal  = 1,
+    kWPAPersonal  = 2,
+    kWPA2Personal = 3,
+    kWPA3Personal = 4,
+};
+
+struct ThreadNetworkInfo
+{
+    uint8_t mDataset[kMaxThreadDatasetLen];
+    uint8_t mDatasetLen;
+};
+
+struct WiFiNetworkInfo
+{
+    uint8_t mSSID[kMaxWiFiSSIDLen + 1];
+    uint8_t mSSIDLen;
+    uint8_t mCredentials[kMaxWiFiCredentialsLen];
+    uint8_t mCredentialsLen;
+    WiFiSecurityType mSecurity;
+};
+
+struct NetworkInfo
+{
+    uint8_t mNetworkID[kMaxNetworkIDLen];
+    uint8_t mNetworkIDLen;
+    uint8_t mEnabled;
+    NetworkType mNetworkType;
+    union NetworkData
+    {
+        ThreadNetworkInfo mThread;
+        WiFiNetworkInfo mWiFi;
+    } mData;
+};
+
+namespace {
+NetworkInfo sNetworks[kMaxNetworks];
+}
+
+void OnAddThreadNetworkCommandCallback(app::Command *, EndpointId, const uint8_t * operationalDataset, uint64_t breadcrumb,
+                                       uint32_t timeoutMs)
+{
+    CHIP_ERROR err                = CHIP_ERROR_NO_MEMORY;
+    size_t operationalDataset_Len = strlen(reinterpret_cast<const char *>(operationalDataset));
```

Comment:
why is the datatype not const char * to start with? Are we guaranteed that this is a null terminated string (and if so, where?)

Reply:
According to the spec, this should be a OCTET_STRING, it should be a uint8_t*. Currently, the octet_string is not handled by current code gen correctly, I may add this support later (as described in the TODOs at the beginning of this file.

Reply:
octet_string should have a length then - null_terminated_octed_string sounds odd.